File: tests/test_scoring.py
#!/usr/bin/env python3
"""
Comprehensive tests for Azul scoring mechanics.
"""

import os
import sys

import pytest

sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from game.game_state import create_game
from game.player_board import PlayerBoard, Wall
from game.tile import Tile, TileColor


class TestWallScoring:
    """Test wall scoring mechanics."""

    def test_single_tile_placement(self):
        """Test scoring for placing a single isolated tile."""
        wall = Wall()

        # Place a single tile - should score 1 point
        points = wall.place_tile(0, TileColor.BLUE)
        assert points == 1, "Single tile should score 1 point"

    def test_horizontal_connections(self):
        """Test scoring for horizontal tile connections."""
        wall = Wall()

        # Place first tile
        wall.place_tile(0, TileColor.BLUE)  # Position (0,0)

        # Place adjacent tile horizontally - should score 2 points
        points = wall.place_tile(0, TileColor.YELLOW)  # Position (0,1)
        assert points == 2, "Two connected tiles should score 2 points"

        # Place third tile in same row - should score 3 points
        points = wall.place_tile(0, TileColor.RED)  # Position (0,2)
        assert points == 3, "Three connected tiles should score 3 points"

    def test_vertical_connections(self):
        """Test scoring for vertical tile connections."""
        wall = Wall()

        # Place first tile
        wall.place_tile(0, TileColor.BLUE)  # Position (0,0)

        # Place adjacent tile vertically - should score 2 points
        points = wall.place_tile(1, TileColor.WHITE)  # Position (1,0)
        assert points == 2, "Two vertically connected tiles should score 2 points"

    def test_cross_connections(self):
        """Test scoring when tile connects both horizontally and vertically."""
        wall = Wall()

        # Create a cross pattern
        wall.place_tile(0, TileColor.BLUE)  # (0,0)
        wall.place_tile(0, TileColor.YELLOW)  # (0,1)
        wall.place_tile(1, TileColor.WHITE)  # (1,0)

        # Place tile that connects both ways
        points = wall.place_tile(1, TileColor.BLUE)  # (1,1)
        assert points == 4, "Cross connection should score 4 points (2+2)"

    def test_complex_scoring_pattern(self):
        """Test complex scoring with multiple connections."""
        wall = Wall()

        # Build a more complex pattern
        wall.place_tile(0, TileColor.BLUE)  # (0,0)
        wall.place_tile(0, TileColor.YELLOW)  # (0,1)
        wall.place_tile(0, TileColor.RED)  # (0,2)
        wall.place_tile(1, TileColor.WHITE)  # (1,0)
        wall.place_tile(2, TileColor.BLACK)  # (2,0)

        # Place tile connecting to both horizontal and vertical lines
        points = wall.place_tile(1, TileColor.BLUE)  # (1,1)
        assert points == 4, "Should score 3 (horizontal) + 1 (vertical) = 4"

    def test_completed_rows_columns_colors(self):
        """Test detection of completed rows, columns, and colors."""
        wall = Wall()

        # Complete first row
        wall.place_tile(0, TileColor.BLUE)
        wall.place_tile(0, TileColor.YELLOW)
        wall.place_tile(0, TileColor.RED)
        wall.place_tile(0, TileColor.BLACK)
        wall.place_tile(0, TileColor.WHITE)

        assert wall.is_row_complete(0), "First row should be complete"
        assert len(wall.get_completed_rows()) == 1

        # Complete first column
        wall.place_tile(1, TileColor.WHITE)
        wall.place_tile(2, TileColor.BLACK)
        wall.place_tile(3, TileColor.RED)
        wall.place_tile(4, TileColor.YELLOW)

        assert wall.is_column_complete(0), "First column should be complete"
        assert len(wall.get_completed_columns()) == 1

        # Check if blue color is complete (need all 5 blue positions)
        wall.place_tile(1, TileColor.BLUE)
        wall.place_tile(2, TileColor.BLUE)
        wall.place_tile(3, TileColor.BLUE)
        wall.place_tile(4, TileColor.BLUE)

        assert wall.is_color_complete(TileColor.BLUE), "Blue color should be complete"
        assert len(wall.get_completed_colors()) == 1


class TestPlayerBoardScoring:
    """Test player board scoring mechanics."""

    def test_pattern_line_completion(self):
        """Test scoring when pattern lines are completed."""
        board = PlayerBoard()

        # Fill pattern line 2 (capacity 3) with blue tiles
        blue_tiles = [Tile(TileColor.BLUE)] * 3
        board.place_tiles_on_pattern_line(2, blue_tiles)

        assert board.pattern_lines[2].is_complete()

        # End round scoring should move tile to wall and score points
        points, discard = board.end_round_scoring()
        assert points == 1, "Single tile placement should score 1 point"
        assert len(discard) == 2, "Should discard 2 excess tiles"
        assert board.wall.filled[2][2], "Blue tile should be on wall at (2,2)"

    def test_floor_line_penalties(self):
        """Test floor line penalty calculations."""
        board = PlayerBoard()

        # Place tiles on floor line
        tiles = [Tile(TileColor.BLUE)] * 7  # Fill all penalty positions
        board.place_tiles_on_floor_line(tiles)

        points, discard = board.end_round_scoring()
        expected_penalty = sum(board.FLOOR_PENALTIES)  # -1-1-2-2-2-3-3 = -14
        assert points == expected_penalty, f"Should have penalty of {expected_penalty}"

    def test_score_cannot_go_negative(self):
        """Test that player score cannot go below 0."""
        board = PlayerBoard()
        board.score = 5

        # Add enough floor penalties to go negative
        tiles = [Tile(TileColor.BLUE)] * 7
        board.place_tiles_on_floor_line(tiles)

        points, _ = board.end_round_scoring()
        assert board.score == 0, "Score should not go below 0"

    def test_first_player_marker_handling(self):
        """Test first player marker on floor line."""
        board = PlayerBoard()

        # Add first player marker to floor line
        first_player_tile = Tile.create_first_player_marker()
        board.place_tiles_on_floor_line([first_player_tile])

        assert board.has_first_player_marker()

        # End round - first player marker should not be discarded
        points, discard = board.end_round_scoring()
        assert not any(tile.is_first_player_marker for tile in discard)

    def test_final_scoring_bonuses(self):
        """Test final scoring bonuses for completed rows/columns/colors."""
        board = PlayerBoard()

        # Manually set up completed patterns for testing
        # Complete first row
        for col in range(5):
            board.wall.filled[0][col] = True

        # Complete first column
        for row in range(5):
            board.wall.filled[row][0] = True

        # Complete blue color (all blue positions)
        blue_positions = [(0, 0), (1, 1), (2, 2), (3, 3), (4, 4)]
        for row, col in blue_positions:
            board.wall.filled[row][col] = True

        initial_score = board.score
        bonus = board.final_scoring()

        expected_bonus = 2 + 7 + 10  # 2 for row, 7 for column, 10 for color
        assert bonus == expected_bonus
        assert board.score == initial_score + expected_bonus


class TestEdgeCaseScoring:
    """Test edge cases in scoring."""

    def test_overflow_tiles_to_floor(self):
        """Test that overflow tiles go to floor line."""
        board = PlayerBoard()

        # Try to place 4 tiles on pattern line with capacity 2
        tiles = [Tile(TileColor.BLUE)] * 4
        board.place_tiles_on_pattern_line(1, tiles)  # Line 1 has capacity 2

        # Since can_add_tiles returns False for 4 tiles on capacity 2 line,
        # all tiles go to floor line
        assert (
            len(board.pattern_lines[1].tiles) == 0
        ), "Pattern line should reject all tiles"
        assert len(board.floor_line) == 4, "All tiles should go to floor"

    def test_invalid_color_placement(self):
        """Test placing wrong color on pattern line."""
        board = PlayerBoard()

        # Place blue tiles on line 0
        blue_tiles = [Tile(TileColor.BLUE)]
        board.place_tiles_on_pattern_line(0, blue_tiles)

        # Try to place red tiles on same line
        red_tiles = [Tile(TileColor.RED)]
        board.place_tiles_on_pattern_line(0, red_tiles)

        # Red tiles should go to floor line
        assert len(board.floor_line) == 1
        assert board.floor_line[0].color == TileColor.RED

    def test_wall_position_already_filled(self):
        """Test placing tile where wall position is already filled."""
        board = PlayerBoard()

        # Manually fill wall position for blue in row 0
        board.wall.filled[0][0] = True

        # Try to place blue tiles on pattern line 0
        blue_tiles = [Tile(TileColor.BLUE)]
        can_place = board.can_place_tiles_on_pattern_line(0, blue_tiles)

        assert (
            not can_place
        ), "Should not be able to place tiles for filled wall position"

    def test_multiple_pattern_lines_complete_same_round(self):
        """Test scoring when multiple pattern lines complete in same round."""
        board = PlayerBoard()

        # Complete multiple pattern lines
        board.place_tiles_on_pattern_line(0, [Tile(TileColor.BLUE)])
        board.place_tiles_on_pattern_line(1, [Tile(TileColor.YELLOW)] * 2)

        points, discard = board.end_round_scoring()

        # Should score for both tiles placed on wall
        assert points == 2, "Should score 1 point for each tile placed"
        assert len(discard) == 1, "Should discard 1 excess tile from line 1"

    def test_empty_pattern_lines_no_scoring(self):
        """Test that empty pattern lines don't affect scoring."""
        board = PlayerBoard()

        # Don't place any tiles
        points, discard = board.end_round_scoring()

        assert points == 0, "No tiles should mean no points"
        assert len(discard) == 0, "No tiles to discard"

    def test_partial_pattern_lines_no_wall_placement(self):
        """Test that incomplete pattern lines don't move to wall."""
        board = PlayerBoard()

        # Partially fill pattern line
        board.place_tiles_on_pattern_line(2, [Tile(TileColor.BLUE)] * 2)  # Capacity 3

        points, discard = board.end_round_scoring()

        assert points == 0, "Incomplete lines should not score"
        assert len(discard) == 0, "No tiles should be discarded"
        assert not board.wall.filled[2][2], "No tile should be placed on wall"


if __name__ == "__main__":
    pytest.main([__file__])

File: game_cpp/test_scoring.cpp
#include "game_state.h"
#include "player_board.h"
#include "tile.h"
#include <iostream>
#include <cassert>
#include <vector>

using namespace azul;

class TestScoring {
public:
    static void test_single_tile_wall_scoring() {
        std::cout << "Testing single tile wall scoring..." << std::endl;
        
        PlayerBoard board;
        
        // Place a single tile on pattern line 0 (capacity 1)
        std::vector<Tile> blue_tiles = {Tile(TileColor::BLUE)};
        board.place_tiles_on_pattern_line(0, blue_tiles);
        
        // Complete the round - should score 1 point for isolated tile
        auto [points, discard] = board.end_round_scoring();
        assert(points == 1);
        
        std::cout << "✓ Single tile wall scoring tests passed" << std::endl;
    }
    
    static void test_horizontal_connections() {
        std::cout << "Testing horizontal connections..." << std::endl;
        
        PlayerBoard board;
        
        // Place tiles to create horizontal connections
        // We need to simulate this by manually setting up the wall
        // or by playing through pattern lines that will create connections
        
        // Place blue tile in row 0
        std::vector<Tile> blue_tiles = {Tile(TileColor::BLUE)};
        board.place_tiles_on_pattern_line(0, blue_tiles);
        board.end_round_scoring();
        
        // Check that tile was placed
        assert(board.wall().is_filled(0, 0)); // Blue goes to position (0,0)
        
        std::cout << "✓ Horizontal connections tests passed" << std::endl;
    }
    
    static void test_floor_line_penalties() {
        std::cout << "Testing floor line penalties..." << std::endl;
        
        PlayerBoard board;
        int initial_score = board.score();
        
        // Place several tiles on floor line to incur penalties
        std::vector<Tile> penalty_tiles;
        for (int i = 0; i < 5; ++i) {
            penalty_tiles.push_back(Tile(TileColor::BLUE));
        }
        board.place_tiles_on_floor_line(penalty_tiles);
        
        // End round - should have penalties
        auto [points, discard] = board.end_round_scoring();
        assert(points < 0); // Should have negative points from penalties
        
        std::cout << "✓ Floor line penalties tests passed" << std::endl;
    }
    
    static void test_score_cannot_go_negative() {
        std::cout << "Testing score cannot go negative..." << std::endl;
        
        PlayerBoard board;
        // Set a low positive score
        // Note: We can't directly set score from outside, so we'll verify behavior
        
        // Fill floor line with many tiles to create large penalty
        std::vector<Tile> many_tiles;
        for (int i = 0; i < 7; ++i) {
            many_tiles.push_back(Tile(TileColor::BLUE));
        }
        board.place_tiles_on_floor_line(many_tiles);
        
        auto [points, discard] = board.end_round_scoring();
        assert(board.score() >= 0); // Score should not go negative
        
        std::cout << "✓ Score cannot go negative tests passed" << std::endl;
    }
    
    static void test_first_player_marker_handling() {
        std::cout << "Testing first player marker handling..." << std::endl;
        
        PlayerBoard board;
        
        // Add first player marker to floor line
        std::vector<Tile> fpm_tiles = {Tile::create_first_player_marker()};
        board.place_tiles_on_floor_line(fpm_tiles);
        
        assert(board.has_first_player_marker());
        
        // End round - first player marker should be handled properly
        auto [points, discard] = board.end_round_scoring();
        
        // First player marker should not be in discard pile
        for (const auto& tile : discard) {
            assert(!tile.is_first_player_marker());
        }
        
        std::cout << "✓ First player marker handling tests passed" << std::endl;
    }
    
    static void test_pattern_line_completion() {
        std::cout << "Testing pattern line completion..." << std::endl;
        
        PlayerBoard board;
        
        // Fill pattern line 2 (capacity 3) completely
        std::vector<Tile> blue_tiles;
        for (int i = 0; i < 3; ++i) {
            blue_tiles.push_back(Tile(TileColor::BLUE));
        }
        board.place_tiles_on_pattern_line(2, blue_tiles);
        
        // Check that pattern line is complete
        assert(board.pattern_lines()[2].is_complete());
        
        // End round scoring
        auto [points, discard] = board.end_round_scoring();
        
        // Should have scored points and discarded excess tiles
        assert(points > 0);
        assert(discard.size() == 2); // Should discard 2 excess tiles
        
        // One tile should be placed on wall
        assert(board.wall().is_filled(2, 2)); // Blue at row 2, col 2
        
        std::cout << "✓ Pattern line completion tests passed" << std::endl;
    }
    
    static void test_multiple_pattern_lines_same_round() {
        std::cout << "Testing multiple pattern lines completion..." << std::endl;
        
        PlayerBoard board;
        
        // Complete multiple pattern lines
        // Line 0 (capacity 1)
        board.place_tiles_on_pattern_line(0, {Tile(TileColor::BLUE)});
        
        // Line 1 (capacity 2)
        std::vector<Tile> red_tiles = {Tile(TileColor::RED), Tile(TileColor::RED)};
        board.place_tiles_on_pattern_line(1, red_tiles);
        
        // Both should be complete
        assert(board.pattern_lines()[0].is_complete());
        assert(board.pattern_lines()[1].is_complete());
        
        // End round scoring
        auto [points, discard] = board.end_round_scoring();
        
        // Should have scored from both lines
        assert(points >= 2); // At least 1 point per tile placed
        
        std::cout << "✓ Multiple pattern lines tests passed" << std::endl;
    }
    
    static void test_empty_pattern_lines_no_scoring() {
        std::cout << "Testing empty pattern lines no scoring..." << std::endl;
        
        PlayerBoard board;
        
        // Don't place any tiles
        
        // End round scoring
        auto [points, discard] = board.end_round_scoring();
        
        // Should not score any points
        assert(points == 0);
        assert(discard.empty());
        
        std::cout << "✓ Empty pattern lines tests passed" << std::endl;
    }
    
    static void test_partial_pattern_lines_no_wall_placement() {
        std::cout << "Testing partial pattern lines no wall placement..." << std::endl;
        
        PlayerBoard board;
        
        // Partially fill pattern line 2 (capacity 3) with only 2 tiles
        std::vector<Tile> blue_tiles = {Tile(TileColor::BLUE), Tile(TileColor::BLUE)};
        board.place_tiles_on_pattern_line(2, blue_tiles);
        
        // Should not be complete
        assert(!board.pattern_lines()[2].is_complete());
        
        // End round scoring
        auto [points, discard] = board.end_round_scoring();
        
        // Should not place tile on wall
        assert(!board.wall().is_filled(2, 2));
        
        // Should not discard the partial tiles
        assert(board.pattern_lines()[2].tiles().size() == 2);
        
        std::cout << "✓ Partial pattern lines tests passed" << std::endl;
    }
    
    static void test_game_end_scoring() {
        std::cout << "Testing game end scoring..." << std::endl;
        
        GameState game(2, 42);
        
        // Play until game ends or we reach a limit
        int actions = 0;
        while (!game.is_game_over() && actions < 1000) {
            auto legal_actions = game.get_legal_actions();
            if (legal_actions.empty()) break;
            game.apply_action(legal_actions[0]);
            actions++;
        }
        
        // Check final scores are reasonable
        auto final_scores = game.get_scores();
        for (int score : final_scores) {
            assert(score >= 0); // Scores should be non-negative
            assert(score < 200); // Reasonable upper bound
        }
        
        std::cout << "✓ Game end scoring tests passed" << std::endl;
    }
    
    static void test_scoring_consistency() {
        std::cout << "Testing scoring consistency..." << std::endl;
        
        // Create two identical games and verify they score the same
        GameState game1(2, 42);
        GameState game2(2, 42);
        
        // Play same actions on both
        for (int i = 0; i < 10 && !game1.is_game_over() && !game2.is_game_over(); ++i) {
            auto actions1 = game1.get_legal_actions();
            auto actions2 = game2.get_legal_actions();
            
            if (actions1.empty() || actions2.empty()) break;
            
            // Should have same legal actions (same seed)
            assert(actions1.size() == actions2.size());
            
            game1.apply_action(actions1[0]);
            game2.apply_action(actions2[0]);
            
            // Should have same scores
            auto scores1 = game1.get_scores();
            auto scores2 = game2.get_scores();
            assert(scores1.size() == scores2.size());
            for (size_t j = 0; j < scores1.size(); ++j) {
                assert(scores1[j] == scores2[j]);
            }
        }
        
        std::cout << "✓ Scoring consistency tests passed" << std::endl;
    }
    
    static void run_all_tests() {
        std::cout << "Running C++ Scoring tests...\n" << std::endl;
        
        try {
            test_single_tile_wall_scoring();
            test_horizontal_connections();
            test_floor_line_penalties();
            test_score_cannot_go_negative();
            test_first_player_marker_handling();
            test_pattern_line_completion();
            test_multiple_pattern_lines_same_round();
            test_empty_pattern_lines_no_scoring();
            test_partial_pattern_lines_no_wall_placement();
            test_game_end_scoring();
            test_scoring_consistency();
            
            std::cout << "\n🎉 All scoring tests passed!" << std::endl;
        } catch (const std::exception& e) {
            std::cout << "\n❌ Test failed: " << e.what() << std::endl;
            throw;
        }
    }
};

int main() {
    TestScoring::run_all_tests();
    return 0;
} 